<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Шаг-1.-Обзор-данных" data-toc-modified-id="Шаг-1.-Обзор-данных-0.1"><span class="toc-item-num">0.1&nbsp;&nbsp;</span>Шаг 1. Обзор данных</a></span></li><li><span><a href="#Шаг-2.1-Заполнение-пропусков" data-toc-modified-id="Шаг-2.1-Заполнение-пропусков-0.2"><span class="toc-item-num">0.2&nbsp;&nbsp;</span>Шаг 2.1 Заполнение пропусков</a></span></li><li><span><a href="#Шаг-2.2-Проверка-данных-на-аномалии-и-исправления." data-toc-modified-id="Шаг-2.2-Проверка-данных-на-аномалии-и-исправления.-0.3"><span class="toc-item-num">0.3&nbsp;&nbsp;</span>Шаг 2.2 Проверка данных на аномалии и исправления.</a></span></li><li><span><a href="#Шаг-2.3.-Изменение-типов-данных." data-toc-modified-id="Шаг-2.3.-Изменение-типов-данных.-0.4"><span class="toc-item-num">0.4&nbsp;&nbsp;</span>Шаг 2.3. Изменение типов данных.</a></span></li><li><span><a href="#Шаг-2.4.-Удаление-дубликатов." data-toc-modified-id="Шаг-2.4.-Удаление-дубликатов.-0.5"><span class="toc-item-num">0.5&nbsp;&nbsp;</span>Шаг 2.4. Удаление дубликатов.</a></span></li><li><span><a href="#Шаг-2.5.-Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма." data-toc-modified-id="Шаг-2.5.-Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма.-0.6"><span class="toc-item-num">0.6&nbsp;&nbsp;</span>Шаг 2.5. Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.</a></span></li><li><span><a href="#Шаг-2.6.-Категоризация-дохода." data-toc-modified-id="Шаг-2.6.-Категоризация-дохода.-0.7"><span class="toc-item-num">0.7&nbsp;&nbsp;</span>Шаг 2.6. Категоризация дохода.</a></span></li><li><span><a href="#Шаг-2.7.-Категоризация-целей-кредита." data-toc-modified-id="Шаг-2.7.-Категоризация-целей-кредита.-0.8"><span class="toc-item-num">0.8&nbsp;&nbsp;</span>Шаг 2.7. Категоризация целей кредита.</a></span></li><li><span><a href="#Ответы-на-вопросы." data-toc-modified-id="Ответы-на-вопросы.-0.9"><span class="toc-item-num">0.9&nbsp;&nbsp;</span>Ответы на вопросы.</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Вопрос-1:-Есть-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок?" data-toc-modified-id="Вопрос-1:-Есть-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок?-0.9.0.1"><span class="toc-item-num">0.9.0.1&nbsp;&nbsp;</span>Вопрос 1: Есть ли зависимость между количеством детей и возвратом кредита в срок?</a></span></li><li><span><a href="#Вывод-1:" data-toc-modified-id="Вывод-1:-0.9.0.2"><span class="toc-item-num">0.9.0.2&nbsp;&nbsp;</span>Вывод 1:</a></span></li><li><span><a href="#Вопрос-2:-Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?" data-toc-modified-id="Вопрос-2:-Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?-0.9.0.3"><span class="toc-item-num">0.9.0.3&nbsp;&nbsp;</span>Вопрос 2: Есть ли зависимость между семейным положением и возвратом кредита в срок?</a></span></li><li><span><a href="#Вывод-2:" data-toc-modified-id="Вывод-2:-0.9.0.4"><span class="toc-item-num">0.9.0.4&nbsp;&nbsp;</span>Вывод 2:</a></span></li><li><span><a href="#Вопрос-3:-Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?" data-toc-modified-id="Вопрос-3:-Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?-0.9.0.5"><span class="toc-item-num">0.9.0.5&nbsp;&nbsp;</span>Вопрос 3: Есть ли зависимость между уровнем дохода и возвратом кредита в срок?</a></span></li><li><span><a href="#Вывод-3:" data-toc-modified-id="Вывод-3:-0.9.0.6"><span class="toc-item-num">0.9.0.6&nbsp;&nbsp;</span>Вывод 3:</a></span></li><li><span><a href="#Вопрос-4:-Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?" data-toc-modified-id="Вопрос-4:-Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?-0.9.0.7"><span class="toc-item-num">0.9.0.7&nbsp;&nbsp;</span>Вопрос 4: Как разные цели кредита влияют на его возврат в срок?</a></span></li><li><span><a href="#Вывод-4:" data-toc-modified-id="Вывод-4:-0.9.0.8"><span class="toc-item-num">0.9.0.8&nbsp;&nbsp;</span>Вывод 4:</a></span></li></ul></li></ul></li></ul></li><li><span><a href="#Общий-вывод:" data-toc-modified-id="Общий-вывод:-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Общий вывод:</a></span></li></ul></div>

# Исследование надежности заемщиков

Заказчик: кредитный отдел банка.

Задача проекта: выяснить, какие факторы (семейное положение, наличие детей) влияют на факт возврата банковского кредита в срок. Полученные выводы будут учитываться при построении модели кредитного скоринга.

Предоставленные данные: статистика о платёжеспособности клиентов в форме таблицы.

Описание данных:
* `children` — количество детей в семье
* `days_employed` — общий трудовой стаж в днях
* `dob_years` — возраст клиента в годах
* `education` — уровень образования клиента
* `education_id` — идентификатор уровня образования
* `family_status` — семейное положение
* `family_status_id` — идентификатор семейного положения
* `gender` — пол клиента
* `income_type` — тип занятости
* `debt` — имел ли задолженность по возврату кредитов
* `total_income` — ежемесячный доход
* `purpose` — цель получения кредита

### Шаг 1. Обзор данных

In [1]:
import pandas as pd
data = pd.read_csv('/datasets/data.csv')
data.head(10)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

In [2]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


В таблице 12 столбцов и 21525 строк.

На первый взгляд видим следующие особенности в данных:
* в столбцах `'days_employed'` и `'total_income'` есть пропущенные значения.
* в столбце `'days_employed'` есть 'странные' значения: не может быть отрицательный стаж.
* также в столбце `'days_employed'` приведены данные типа float, хотя количество дней может быть представлено целыми числами.

### Шаг 2.1 Заполнение пропусков

Находим, сколько пропущенных значений в столбцах `'days_employed'` и `'total_income'` и выясняем, какую долю это занимает от общего количества значений.

In [3]:
print(data['days_employed'].isna().sum())
print(data['total_income'].isna().sum())
print(data['days_employed'].isna().sum() / len(data['days_employed']))

2174
2174
0.10099883855981417


В обоих столбцах одинаковое количество пропущенных значений. Учитывая, что в столбцах одинаковое также общее количество значений, можно сказать, что и в одном, и в другом случае **пропущено 10%**.

Можно изучить, есть ли особенности у строк, в которых пропущены данные этих столбцов.

In [4]:
display(data[data['days_employed'].isna()].head(10))

children  days_employed  dob_years education  education_id  \
12         0            NaN         65   среднее             1   
26         0            NaN         41   среднее             1   
29         0            NaN         63   среднее             1   
41         0            NaN         50   среднее             1   
55         0            NaN         54   среднее             1   
65         0            NaN         21   среднее             1   
67         0            NaN         52    высшее             0   
72         1            NaN         32    высшее             0   
82         2            NaN         50    высшее             0   
83         0            NaN         52   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   
65  Не женат / не замужем                 4      M    компаньон     0   
67        женат / замужем                 0      F    пенсионер     0   
72        женат / замужем                 0      M  госслужащий     0   
82        женат / замужем                 0      F    сотрудник     0   
83        женат / замужем                 0      M    сотрудник     0   

    total_income                                purpose  
12           NaN                        сыграть свадьбу  
26           NaN                            образование  
29           NaN       строительство жилой недвижимости  
41           NaN       сделка с подержанным автомобилем  
55           NaN                        сыграть свадьбу  
65           NaN  операции с коммерческой недвижимостью  
67           NaN                покупка жилья для семьи  
72           NaN  операции с коммерческой недвижимостью  
82           NaN                                  жилье  
83           NaN                                  жилье

Каких-либо закономерностей не видно, за исключением того, что в строках с пропуском в столбце `'days_employed'` также нет данных и в столбце `'total_income'`. Вряд ли пропуски связаны с какой-то тенденцией: они могли появиться из-за человеческой ошибки.

Заменим пропуски в столбце `'total_income'` на **медианное значение**: его лучше использовать в тех случах, когда речь идёт об измерении доходов и других показателей с большой вариативностью, так как минимальные и максимальные значения могут очень сильно влиять на среднее. 

In [5]:
total_income_median = data['total_income'].median()
print(total_income_median)
# заменяем полученным медианным значением пропуски в данных
data['total_income'] = data['total_income'].fillna(total_income_median)
# проверяем, не осталось ли пропусков
print(data['total_income'].isna().sum())

145017.93753253992
0


### Шаг 2.2 Проверка данных на аномалии и исправления.

Необходимо исправить аномалии в столбце `'days_employed'`, чтобы избавиться от отрицательных значений. Возьмём все данные по модулю.

In [6]:
data['days_employed'] = data['days_employed'].abs()
# проверяем, исчезли ли отрицательные значения
print(data['days_employed'].min())
print(data['days_employed'].max())


24.14163324048118
401755.40047533


Обнаружено странное максимальное значение. Получается, что клиент работал 1100 лет. Можно уточнить, сколько в таблице подобных аномальных значений.

In [7]:
# Примем, что трудовой стаж человека в возрасте 75 лет не может быть более 100 лет 
# (даже с учётом "северной надбавки") или 36500 дней

employees_of_millenia = data.loc[data['days_employed'] > 36500]
print(len(employees_of_millenia['days_employed']))

# возможно, проблема в конкретной профессиональной группе
print(employees_of_millenia.groupby('income_type')['dob_years'].count())

# сравним с количеством клиентов из этих профессиональных групп в общей таблице
data.groupby('income_type')['dob_years'].count()

3445
income_type
безработный       2
пенсионер      3443
Name: dob_years, dtype: int64


income_type
безработный            2
в декрете              1
госслужащий         1459
компаньон           5085
пенсионер           3856
предприниматель        2
сотрудник          11119
студент                1
Name: dob_years, dtype: int64

У обоих безработных и многих пенсионеров стаж указан неправильно. Можно нормировать экстремальные показатели, разделив их на 20 (тогда максимальный стаж будет 55 лет, что похоже на правду. Такой подход здесь возможно применить, так как столбец стажа не понадобится для целей исследования - в целом более оптимальным было бы прояснить ситуацию у тех, кто собирал данные.

In [8]:
for item in data['days_employed']: 
    if item > (55 * 365):
        data['days_employed'] = data['days_employed'].replace(item, item / 20)

print(data['days_employed'].max())

20087.7700237665


Теперь можно заменить пропущенные значения в этом столбце на медианное значение.

In [9]:
days_employed_median = data['days_employed'].median()
print(days_employed_median)
# заменяем полученным медианным значением пропуски в данных
data['days_employed'] = data['days_employed'].fillna(days_employed_median)
# проверяем, не осталось ли пропусков
print(data['days_employed'].isna().sum())

2194.220566878695
0


**Проверим на аномалии другие столбцы с количественныеми переменными.**

In [10]:
print(data['children'].sort_values().unique())
# в столбце 'children' видим аномалию - отрицательным количество детей быть не может

# посмотрим более детально
print(data['children'].value_counts())


[-1  0  1  2  3  4  5 20]
 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64


In [11]:
print(data['dob_years'].min())
print(data['dob_years'].max())
# видим значение '0', что выглядит подозрительно - не может быть такого клиента банка

# посмотрим, сколько аномальных значений в этом столбце в целом
print(data['dob_years'].sort_values().unique())
# кроме '0' аномальных значений нет, значит, это значение символизирует пропуск данных

# посмотрим, сколько раз данные о возрасте были пропущены
print(len(data[data['dob_years'] == 0]))
print()

0
75
[ 0 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41
 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65
 66 67 68 69 70 71 72 73 74 75]
101



In [12]:
print(data['debt'].unique())
# '0' означает, что задолженности у клиента не было, '1' - что была

[0 1]


**Аномальные значения есть в столбцах**: 
* `'children'` - если основываться на данных этого столбца, у некоторых клиентов отрицательное количество детей. Видимо, имело место опечатка. С другой стороны, есть несколько наблюдений, в которых значение соответствует 20. По идее, причину происхождения этой "ошибки" нужно было бы прояснить с теми, кто собирает данные. Возможно, таким образом записывается многодетность, а возможно - там лишний ноль. Но так как этих значений немного, они вряд ли повлияют на выводы исследования.
* `'dob_years'` - по всей видимости, '0' означает пропуск данных. Таких случаев довольно мало: около 4%, но всё же их целесообразно изменить.

Исправим проблемы. В столбце `'children'` приведём все значения к натуральным, а в столбце `'dob_years'` заменим нули на средний возраст клиентов. В данном случае оправдано использование среднего, так как разброс данных по возрасту небольшой, нет экстремальных выбросов, которые бы влияли на среднее.

In [13]:
data['children'] = data['children'].abs()

dob_years_mean = data['dob_years'].mean().astype(int)
# лучше привести среднее к типу integer, характерному для остальных данных в этом столбце

# заменяем нули в столбце на полученное среднее значение
data['dob_years'] = data['dob_years'].replace(0, dob_years_mean)



### Шаг 2.3. Изменение типов данных.

Остаётся проблема типа данных: нужно заменить тип данных float на int в столбце `'days_employed'`.

In [14]:
data['days_employed'] = data['days_employed'].astype(int)

# проверяем, какие типы данных в столбцах
data.dtypes


children              int64
days_employed         int64
dob_years             int64
education            object
education_id          int64
family_status        object
family_status_id      int64
gender               object
income_type          object
debt                  int64
total_income        float64
purpose              object
dtype: object

### Шаг 2.4. Удаление дубликатов.

Сначала необходимо выяснить, есть ли в таблице строки-дубликаты и сколько их.

In [15]:
data.duplicated().sum()

54

Избавимся от дубликатов.

In [16]:
data = data.drop_duplicates().reset_index(drop=True)

# проверим, остались ли дубликаты
data.duplicated().sum()

0

Проверим, есть ли **неявные дубликаты** во всех столбцах, содержащих категориальные значения: `'education'`, `'family_status'`, `'gender'`, `'income_type'` и `'purpose'`.

**Столбец `'education'`**

In [17]:
data['education'].sort_values().unique()

array(['ВЫСШЕЕ', 'Высшее', 'НАЧАЛЬНОЕ', 'НЕОКОНЧЕННОЕ ВЫСШЕЕ',
       'Начальное', 'Неоконченное высшее', 'СРЕДНЕЕ', 'Среднее',
       'УЧЕНАЯ СТЕПЕНЬ', 'Ученая степень', 'высшее', 'начальное',
       'неоконченное высшее', 'среднее', 'ученая степень'], dtype=object)

Заметно, что все дубликаты в этом столбце связаны с использованием разных регистров.

Вероятно, такие дубликаты появились из-за "человеческой ошибки": например, разные сотрудники могли по-разному заполнять данные при работе с клиентами.

Устраним проблему приведением всех данных к нижнему регистру.

In [18]:
data['education'] = data['education'].str.lower()

data['education'].sort_values().unique()

array(['высшее', 'начальное', 'неоконченное высшее', 'среднее',
       'ученая степень'], dtype=object)

**Столбцы `'family_status'`, `'gender'` и `'income_type'`**

In [19]:
data['family_status'].sort_values().unique()

array(['Не женат / не замужем', 'в разводе', 'вдовец / вдова',
       'гражданский брак', 'женат / замужем'], dtype=object)

In [20]:
print(data['gender'].sort_values().unique())
# получили странный вывод
print(data['gender'].value_counts())
data.loc[data['gender'] == 'XNA']

['F' 'M' 'XNA']
F      14189
M       7281
XNA        1
Name: gender, dtype: int64


children  days_employed  dob_years            education  education_id  \
10690         0           2358         24  неоконченное высшее             2   

          family_status  family_status_id gender income_type  debt  \
10690  гражданский брак                 1    XNA   компаньон     0   

        total_income               purpose  
10690  203905.157261  покупка недвижимости

Пол не указан всего в одном случае из 21500 наблюдений. Вряд ли это повлияет на результаты исследования.

In [21]:
data['income_type'].sort_values().unique()

array(['безработный', 'в декрете', 'госслужащий', 'компаньон',
       'пенсионер', 'предприниматель', 'сотрудник', 'студент'],
      dtype=object)

Во всех трёх столбцах нет дубликатов.

**Столбец `'purpose'`** 

In [22]:
data['purpose'].sort_values().unique()

array(['автомобили', 'автомобиль', 'высшее образование',
       'дополнительное образование', 'жилье',
       'заняться высшим образованием', 'заняться образованием',
       'на покупку автомобиля', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля', 'на проведение свадьбы',
       'недвижимость', 'образование', 'операции с жильем',
       'операции с коммерческой недвижимостью',
       'операции с недвижимостью', 'операции со своей недвижимостью',
       'покупка жилой недвижимости', 'покупка жилья',
       'покупка жилья для сдачи', 'покупка жилья для семьи',
       'покупка коммерческой недвижимости', 'покупка недвижимости',
       'покупка своего жилья', 'получение высшего образования',
       'получение дополнительного образования', 'получение образования',
       'приобретение автомобиля', 'профильное образование',
       'ремонт жилью', 'свадьба', 'свой автомобиль',
       'сделка с автомобилем', 'сделка с подержанным автомобилем',
       'строительство 

In [23]:
data['purpose'].value_counts()

свадьба                                   793
на проведение свадьбы                     773
сыграть свадьбу                           769
операции с недвижимостью                  675
покупка коммерческой недвижимости         662
покупка жилья для сдачи                   652
операции с жильем                         652
операции с коммерческой недвижимостью     650
жилье                                     646
покупка жилья                             646
покупка жилья для семьи                   638
строительство собственной недвижимости    635
недвижимость                              633
операции со своей недвижимостью           627
строительство жилой недвижимости          625
покупка недвижимости                      621
покупка своего жилья                      620
строительство недвижимости                619
ремонт жилью                              607
покупка жилой недвижимости                606
на покупку своего автомобиля              505
заняться высшим образованием      

Дубликатов в этом столбце очень много, но **нельзя сказать, что они полностью совпадают по смыслу**. Например, "автомобили" и "на покупку автомобиля": вполне может быть, что первый клиент взял кредит на ремонт машины, а не на покупку. Нельзя также сказать, что одни обозначения целей встречаются в разы чаще, чем другие. 

Видимо, этот столбец заполнялся разными сотрудниками со слов клиентов.

Таким образом, устранять дубликаты в этом столбце нецелесообразно - лучше их категоризировать для целей исследования.

### Шаг 2.5. Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.

Создаём два новых "словаря", чтобы не "загромождать" исходную таблицу лишними столбцами с текстом.

* В словаре **`education_dict`** будут содержаться данные о том, какому индексу соответствует какой уровень образования клиента.
* В словаре **`family_status_dict`** будут данные о соответствии индекса семейному положению клиента.

In [24]:
education_dict = data[['education', 'education_id']]
education_dict = education_dict.drop_duplicates().reset_index(drop=True)

education_dict.head()

education  education_id
0               высшее             0
1              среднее             1
2  неоконченное высшее             2
3            начальное             3
4       ученая степень             4

In [25]:
family_status_dict = data[['family_status', 'family_status_id']]
family_status_dict = family_status_dict.drop_duplicates().reset_index(drop=True)

family_status_dict.head()

family_status  family_status_id
0        женат / замужем                 0
1       гражданский брак                 1
2         вдовец / вдова                 2
3              в разводе                 3
4  Не женат / не замужем                 4

Удалим из основного датафрейма столбцы с семейным положением и образованием.

In [26]:
data = data.drop(columns=['family_status', 'education'])
# проверим, как сработал метод drop()
data.head()

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437         42             0                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0          17013         53             1                 1      F   

  income_type  debt   total_income                     purpose  
0   сотрудник     0  253875.639453               покупка жилья  
1   сотрудник     0  112080.014102     приобретение автомобиля  
2   сотрудник     0  145885.952297               покупка жилья  
3   сотрудник     0  267628.550329  дополнительное образование  
4   пенсионер     0  158616.077870             сыграть свадьбу

### Шаг 2.6. Категоризация дохода.

Задачу категоризации доходов клиентов решим с помощью создания нового столбца `'total_income_category'`, в котором доходы будут распределены по категориям следующим образом:
* 0–30000 — 'E';
* 30001–50000 — 'D';
* 50001–200000 — 'C';
* 200001–1000000 — 'B';
* 1000001 и выше — 'A'.

In [27]:
# создадим функцию income_group
def income_group(income):
    if income <= 30000:
        return 'E'
    elif income <= 50000:
        return 'D'
    elif income <= 200000:
        return 'C'
    elif income <= 1000000:
        return 'B'
    return 'A'
# проверим работу функции
# print(income_group(1100000))
# функция работает

# применим функцию к датафрейму
data['total_income_category'] = data['total_income'].apply(income_group)
data.head()

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437         42             0                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0          17013         53             1                 1      F   

  income_type  debt   total_income                     purpose  \
0   сотрудник     0  253875.639453               покупка жилья   
1   сотрудник     0  112080.014102     приобретение автомобиля   
2   сотрудник     0  145885.952297               покупка жилья   
3   сотрудник     0  267628.550329  дополнительное образование   
4   пенсионер     0  158616.077870             сыграть свадьбу   

  total_income_category  
0                     B  
1                     C  
2                     C  
3                     B  
4                     C

### Шаг 2.7. Категоризация целей кредита.

Ещё раз выведем набор переменных, описывающих цели кредитов. Известно, что категоризировать их необходимо на 4 группы:
* 'операции с автомобилем',
* 'операции с недвижимостью',
* 'проведение свадьбы',
* 'получение образования'.

In [28]:
data['purpose'].sort_values().unique()

array(['автомобили', 'автомобиль', 'высшее образование',
       'дополнительное образование', 'жилье',
       'заняться высшим образованием', 'заняться образованием',
       'на покупку автомобиля', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля', 'на проведение свадьбы',
       'недвижимость', 'образование', 'операции с жильем',
       'операции с коммерческой недвижимостью',
       'операции с недвижимостью', 'операции со своей недвижимостью',
       'покупка жилой недвижимости', 'покупка жилья',
       'покупка жилья для сдачи', 'покупка жилья для семьи',
       'покупка коммерческой недвижимости', 'покупка недвижимости',
       'покупка своего жилья', 'получение высшего образования',
       'получение дополнительного образования', 'получение образования',
       'приобретение автомобиля', 'профильное образование',
       'ремонт жилью', 'свадьба', 'свой автомобиль',
       'сделка с автомобилем', 'сделка с подержанным автомобилем',
       'строительство 

Заметно, что в названии каждой из групп, как и в разнообразии целей кредита есть **ключевые слова**, которые помогут распределить конкретные цели в группы. Напишем функцию, которая позволит это сделать и создадим столбец `'purpose_category'`.

In [29]:
def purpose_group(purpose):
    if 'авто' in purpose:
        return 'операции с автомобилем'
    elif ('недвиж' in purpose) | ('жиль' in purpose):
        return 'операции с недвижимостью'
    elif 'свадьб' in purpose:
        return 'проведение свадьбы'
    elif 'образован' in purpose:
        return 'получение образования'
    return 'категория не определена'
# проверим работу функции
# print(purpose_group('дополнительное образование'))
# функция работает

# С помощью функции создаём столбец
data['purpose_category'] = data['purpose'].apply(purpose_group)

# проверяем, появился ли новый столбец
display(data.head())

# посмотрим, есть ли в созданном столбце ячейки, где "категория не определена"
print(data.loc[data['purpose_category'] == 'категория не определена'])

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437         42             0                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0          17013         53             1                 1      F   

  income_type  debt   total_income                     purpose  \
0   сотрудник     0  253875.639453               покупка жилья   
1   сотрудник     0  112080.014102     приобретение автомобиля   
2   сотрудник     0  145885.952297               покупка жилья   
3   сотрудник     0  267628.550329  дополнительное образование   
4   пенсионер     0  158616.077870             сыграть свадьбу   

  total_income_category          purpose_category  
0                     B  операции с недвижимостью  
1                     C    операции с автомобилем  
2                     C  операции с недвижимостью  
3                     B     получение образования  
4                     C        проведение свадьбы

Empty DataFrame
Columns: [children, days_employed, dob_years, education_id, family_status_id, gender, income_type, debt, total_income, purpose, total_income_category, purpose_category]
Index: []


Таким образом, все варианты описания целей кредита удалось категоризировать.

Данные готовы к анализу.

### Ответы на вопросы.

##### Вопрос 1: Есть ли зависимость между количеством детей и возвратом кредита в срок?

Прежде всего, для сравнения нужно понять, сколько в среднем должников среди всех клиентов.

In [30]:
print(data['debt'].sum())
print(data['debt'].mean())

1741
0.08108611615667645


Таким образом, в среднем должников около 8,1%.

In [31]:
# для ответа на вопрос проще всего сгруппировать клиентов по количеству детей 
# и найти в каждой группе среднее количество должников

print(data.groupby('children')['debt'].agg(['count', 'sum', 'mean']))


          count   sum      mean
children                       
0         14107  1063  0.075353
1          4856   445  0.091639
2          2052   194  0.094542
3           330    27  0.081818
4            41     4  0.097561
5             9     0  0.000000
20           76     8  0.105263


##### Вывод 1:

Больше всего должников в условной группе "очень многодетных" - около 10%, что заметно выше, чем в среднем по всем клиентам. Правда, в этой группе мало наблюдений - недостаточно для того, чтобы говорить о тенденции.

Значения выше среднего также наблюдаются у клиентов с 4, 2 детьми и 1 ребёнком.

Среди клиентов с 3 детьми количество должников такое же, как в среднем по группе, но эта категория клиентов, опять же, малочисленна.

Бездетные клиенты являются наиболее дисциплинированными заёмщиками: среди них доля должников меньше, чем в среднем. Нет должников среди группы клиентов с 5 детьми, однако в этой группе всего 9 человек - сделать выводы о тенденции не представляется возможным.

Таким образом, можно говорить о том, что наличие детей у клиентов как таковое является фактором риска для банка, так как **среди клиентов с детьми доля должников выше, чем в среднем**. Очевидно, причина этого в том, что клиенты с детьми часть своих доходов должны тратить на содержание детей, у них более вероятно возникновение форс-мажоров, предполагающих внезапные расходы. Другими словами, **их финансовая ситуация является менее предсказуемой, что может повлечь просрочку по кредиту**. И чем больше детей, тем непредсказуемость выше.

<div class="alert alert-success">
    
<b> Комментарий от ревьюера ✔️</b>
    
Хороший комментарий. Здорово, что обратила внимание не только на долю клиентов с задолженностью, но и на размеры групп
</div>

##### Вопрос 2: Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [32]:
print(data.groupby('family_status_id')['debt'].mean().sort_values(ascending=False))

# формируем таблицу для оценки абсолютных значений
family_table = data.pivot_table(index='family_status_id', columns='debt', 
                                  values='dob_years', aggfunc='count')

# совместим со "словарём"
family_table = family_table.merge(family_status_dict, on='family_status_id', how='left')
family_table.drop('family_status_id', axis=1)

family_status_id
4    0.097509
1    0.093202
0    0.075421
3    0.071130
2    0.065693
Name: debt, dtype: float64


0    1          family_status
0  11413  931        женат / замужем
1   3775  388       гражданский брак
2    896   63         вдовец / вдова
3   1110   85              в разводе
4   2536  274  Не женат / не замужем

##### Вывод 2:

Доля должников заметно больше (и выше, чем в среднем) среди клиентов, не состоящих в браке, а также среди тех, кто состоит в гражданском браке.

Среди клиентов, находящихся в зарегистрированном браке, доля должников приближена к средней по выборке.

Наиболее дисциплинированные заёмщики, среди которых должники встречаются заметно реже, чем в целом по выборке, - это клиенты, находящиеся в разводе, и вдовцы/вдовы. 

Возникает гипотеза, что *более дисциплинированными могут быть более пожилые заёмщики*. Она нуждается в проверке:

In [33]:
age_family_table = data.pivot_table(index='family_status_id', columns='debt', 
                                  values='dob_years', aggfunc='mean')

# более наглядная таблица
age_family_table = age_family_table.merge(family_status_dict, on='family_status_id', how='left')
age_family_table.drop('family_status_id', axis=1)

0          1          family_status
0  43.951809  40.873255        женат / замужем
1  42.552583  40.105670       гражданский брак
2  56.767857  56.158730         вдовец / вдова
3  46.026126  43.941176              в разводе
4  38.957808  35.375912  Не женат / не замужем

Действительно, во всех группах средний возраст тех, кто допустил просрочку, чуть ниже соответствующего показателя дисциплинированных должников.

Также заметно, что средний возраст двух наиболее дисциплинированных категорий - вдов/вдовцов и клиентов, находящихся в разводе, заметно выше, чем средний возраст остальных групп.

##### Вопрос 3: Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [34]:
data.groupby('total_income_category')['debt'].mean().sort_values(ascending=False)

total_income_category
E    0.090909
C    0.084825
A    0.080000
B    0.070621
D    0.060000
Name: debt, dtype: float64

In [35]:
income_table = data.pivot_table(index=['total_income_category', 'income_type'], columns='debt', 
                                values='dob_years', aggfunc='count', fill_value=0)
# заменили значения NaN, возникающие, если нет данных, на 0

income_table

debt                                      0    1
total_income_category income_type               
A                     компаньон          13    2
                      сотрудник          10    0
B                     безработный         1    0
                      госслужащий       351   18
                      компаньон        1631  110
                      пенсионер         521   31
                      предприниматель     1    0
                      сотрудник        2180  197
C                     безработный         0    1
                      в декрете           0    1
                      госслужащий       996   67
                      компаньон        3035  263
                      пенсионер        2929  176
                      предприниматель     1    0
                      сотрудник        7711  852
                      студент             1    0
D                     госслужащий        23    1
                      компаньон          24    1
                      пенсионер         158    7
                      сотрудник         124   12
E                     госслужащий         1    0
                      компаньон           1    0
                      пенсионер          13    2
                      сотрудник           5    0

##### Вывод 3:

Наименее дисциплинированными заёмщиками оказались клиенты из группы с самыми низкими доходами. Тем не менее, в этой группе очень мало наблюдений для того, чтобы можно было делать обоснованные выводы.

Предпоследняя по доходам группа значительно более многочисленна, при этом она оказалась наиболее дисциплинированной. 

Ниже среднего значения просрочка также в группе 'B' (более 200 тыс. и до 1 млн.) - по всей видимости, эта группа имеет меньше финансовых рисков и берёт "посильные" ссуды.

В группах с самым высоким уровнем доходов (она также невелика) и со средними доходами ('С') клиенты возвращают кредиты примерно так же, как в среднем по выборке. Необходимо заметить, что в группе 'C' наиболее "рискованной" категорией являются сотрудники организаций.

Сравнительно большая доля задолженностей в группе с самыми низкими доходами вполне объяснима: в такой ситуации у заёмщика множество рисков для финансовой стабильности. У остальных групп повышенные риски могут быть связаны с суммой кредита.

##### Вопрос 4: Как разные цели кредита влияют на его возврат в срок?

In [36]:
data.groupby('purpose_category')['debt'].mean().sort_values(ascending=False)

purpose_category
операции с автомобилем      0.093547
получение образования       0.092177
проведение свадьбы          0.079657
операции с недвижимостью    0.072314
Name: debt, dtype: float64

In [37]:
category_table = data.pivot_table(index=['purpose_category', 'total_income_category'], columns='debt', 
                                values='total_income', aggfunc='count', fill_value=0)
# заменили значения NaN, возникающие, если нет данных, на 0
category_table

debt                                               0    1
purpose_category         total_income_category           
операции с автомобилем   A                         2    0
                         B                       945   86
                         C                      2894  307
                         D                        62   10
                         E                         2    0
операции с недвижимостью A                        16    1
                         B                      2396  172
                         C                      7449  601
                         D                       162    6
                         E                         9    2
получение образования    A                         3    1
                         B                       827   70
                         C                      2735  297
                         D                        74    2
                         E                         5    0
проведение свадьбы       A                         2    0
                         B                       517   28
                         C                      1595  155
                         D                        31    3
                         E                         4    0

##### Вывод 4:

Более высокая доля просроченных займов - в категориях операций с автомобилем и получения образования. Среди тех, кто берёт кредит на автомобиль, худшая ситуация с просрочкой у клиентов со средними и невысокими доходами. Та же ситуация и с кредитами на образование: более 10% клиентов со средними доходами не выплачивают их в срок.

Наиболее вероятен возврат в срок кредитов на операции с недвижимостью (хотя это не касается заёмщиков из двух групп с доходами ниже среднего). Как представляется, надёжность в данном случае обеспечивается доскональным контролем кредитной истории заёмщика и рядом других мер со стороны банка перед выдачей крупного займа.

Что касается кредитов на проведение свадьбы, то возникновение долга по ним происходит лишь чуть реже, чем в среднем.

*NB Поскольку речь зашла об образовании, интересно проверить, зависит ли надёжность заёмщика от образования. Оказывается, наличие у клиента высшего образования повышает вероятность возврата кредита в срок самым заметным образом из всех факторов, исследованных ранее.*

In [38]:
display(data.groupby('education_id')['debt'].mean().sort_values(ascending=False))

education_table = data.pivot_table(index='education_id', columns='debt', 
                                  values='dob_years', aggfunc='count')
education_table = education_table.merge(education_dict, on='education_id', how='left')
education_table.drop('education_id', axis=1)

education_id
3    0.109929
2    0.091398
1    0.089808
0    0.052942
4    0.000000
Name: debt, dtype: float64

0       1            education
0   4973.0   278.0               высшее
1  13824.0  1364.0              среднее
2    676.0    68.0  неоконченное высшее
3    251.0    31.0            начальное
4      6.0     NaN       ученая степень

## Общий вывод:

Можно говорить о том, что **высокие риски возникновения задолженности** связаны со следующими факторами:
* наличие у клиента детей - в особенности, больше 1,
* клиент в молодом или среднем возрасте и не состоит и никогда не состоял в официально зарегистрированном браке,
* уровень дохода клиента либо очень низкий (до 30 тыс.), либо он является сотрудником организации (не госслужащим) и имеет доход свыше 50 тыс. но не более 200 тыс.
* клиент берёт кредит на получение образования или операции с автомобилем.

**Снижению рисков просрочки** способствуют такие факторы, как:
* бездетность клиента,
* статус вдовца/вдовы или пребывание в разводе,
* средне-высокие или средне-низкие доходы (либо от 200 тыс. до 1 млн., либо от 30 до 50 тыс.),
* получение кредита на операции с недвижимостью,
* наличие у клиента высшего образования.